packages/pages/src/vite-plugin/modules/plugin.ts

```diff
@@ -0,0 +1,173 @@
+      const filepath = path.resolve(f);
+      const moduleName = getModuleName(filepath);
+      filepaths[moduleName] = filepath;
+    });
+
+  const logger = createLogger();
+  const loggerInfo = logger.info;
+
+  for (const [moduleName, modulePath] of Object.entries(filepaths)) {
+    const index = addExtraModuleCode(modulePath, moduleName);
+    logger.info = (msg, options) => {
+      if (msg.includes("building for production")) {
+        loggerInfo(pc.green(`\nBuilding ${moduleName} module...`));
+        return;
+      }
+      loggerInfo(msg, options);
+    };
+
+    await build({
+      customLogger: logger,
+      configFile: false,
+      envDir: envVarConfig.envVarDir,
+      envPrefix: envVarConfig.envVarPrefix,
+      resolve: {
+        conditions: ["worker", "webworker"],
+      },
+      publicDir: false,
+      css: {
+        postcss: path.join(
+          rootFolders.source,
+          subfolders.modules,
+          `${moduleName}/postcss.config.cjs`
+        ),
+      },
+      build: {
+        emptyOutDir: false,
+        outDir: outdir,
+        minify: true,
+        rollupOptions: {
+          input: modulePath,
+          output: {
+            format: "umd",
+            entryFileNames: `${moduleName}.umd.js`,
+          },
+        },
+        reportCompressedSize: false,
+      },
+      define: processEnvVariables(envVarConfig.envVarPrefix),
+      plugins: [
+        nodePolyfills({
+          globals: {
+            Buffer: "build",
+            global: "build",
+            process: "build",
+          },
+        }),
+      ],
+    });
+    removeAddedModuleCode(modulePath, index);
+  }
+};
+
+const shouldBundleModules = (projectStructure: ProjectStructure) => {
+  const { rootFolders, subfolders } = projectStructure.config;
+  return fs.existsSync(path.join(rootFolders.source, subfolders.modules));
+};
+
+/**
+ *
+ * @param modulePath
+ * @returns name of module if set by user, else uses file name
+ */
+const getModuleName = (modulePath: string): string => {
+  const sfp = new SourceFileParser(modulePath, createTsMorphProject());
+  const { name } = path.parse(modulePath);
+  return (
+    sfp
+      .getVariablePropertyByType("ModuleConfig", "name")
+      ?.replace(/['"`]/g, "") ?? name
+  );
+};
+
+/**
+ * Adds custom code to module such that it works for user when bundled into umd.js.
+ *
+ * @param modulePath
+ * @param name
+ * @returns number of added index
+ */
+const addExtraModuleCode = (modulePath: string, name: string): number => {
+  const sfp = new SourceFileParser(modulePath, createTsMorphProject());
+  const declaration = sfp.getVariableDeclarationByType("Module");
+  if (declaration === undefined) {
+    throw new Error(`Cannot find variable Module in ${modulePath}`);
+  }
+  const moduleName = declaration.getName();
+  const insertIndex = sfp.getEndPos();
+  const afterInsertIndex = sfp.insertStatement(
+    wrappedCode(moduleName, name),
+    insertIndex
+  );
+  sfp.addReactImports();
```

Comment:
Do we actually need this? Is the react stuff because of the ReactDOM.render we're using? Or is it just if the user doesn't add it to their template? If the latter I don't think we need to do it for them automatically.

Reply:
>  Is the react stuff because of the ReactDOM.render we're using?


Yes! We do need it for that.